node/src/components/consensus/era_supervisor.rs

```diff
@@ -181,47 +180,32 @@ where
             era_where_we_joined: current_era,
         };
 
-        let bonded_eras = era_supervisor.bonded_eras();
-        let era_ids: Vec<EraId> = era_supervisor
-            .iter_past(current_era, bonded_eras.saturating_mul(3))
-            .collect();
-
         // Collect the information needed to initialize all recent eras.
-        // TODO: Merge these into one
-        let mut key_blocks = BTreeMap::new();
-        let mut booking_blocks = BTreeMap::new();
-        for era_id in era_ids {
-            let minimum_era = era_supervisor
-                .protocol_config
-                .last_emergency_restart
-                .unwrap_or_else(|| EraId::from(0u64));
-            let key_block_era_id = era_id.saturating_sub(1);
-            let block_header = storage
-                .read_switch_block_header_by_era_id(key_block_era_id)?
+        let auction_delay = era_supervisor.protocol_config.auction_delay;
+        let last_activation_point = era_supervisor.protocol_config.last_activation_point;
+        let mut switch_blocks = Vec::new();
+        // We need to initialize current_era, current_era - 1 and current_era - 2. To initialize an
+        // era, all switch blocks between its booking block and its key block are required. The
+        // booking block for era N is in N - auction_delay - 1, and the key block in N - 1.
+        // So we need all switch blocks between (including) current_era - 2 - auction_delay - 1
+        // and (excluding) current_era. However, we never use any block from before the last
+        // activation point.
+        let earliest_era =
+            last_activation_point.max(current_era.saturating_sub(auction_delay).saturating_sub(3));
+        for era_id in (earliest_era.value()..current_era.value()).map(EraId::from) {
+            let switch_block = storage
+                .read_switch_block_header_by_era_id(era_id)?
                 .ok_or_else(|| anyhow::Error::msg("No such switch block"))?;
```

Comment:
Can you add an `era_id` to the error message please?

Reply:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/273073a47ece41dfdb75419b614c945b9fb11868.